WIP Qt6 port

gfgit opened the pull request:
Depends on:

- [lxqt-build-tools#77](https://github.com/lxqt/lxqt-build-tools/pull/77)
- [liblxqt/pull/337](https://github.com/lxqt/liblxqt#337)
- [lxqt-globalkeys#278](https://github.com/lxqt/lxqt-globalkeys/pull/278)
- [lxqt-menu-data#26](https://github.com/lxqt/lxqt-menu-data/pull/26)
- [libqtxdg#272](https://github.com/lxqt/libqtxdg/pull/272)

AzumaHazuki commented:
This is amazing! Where did you learn all this coding?

gfgit commented:
Hi! Thanks. Well it does not work yet, I need more time to properly fix compile errors.
Anyway it's not that difficult, I've just read Qt online documentation and made some experiments on small test programs until I get the results I want. The key for me is having fun while coding, otherwise is not worth it.

stefonarch commented:
Looks like only fancymenu and few other plugins build atm. Made a list [here](https://paste.debian.net/1307157/).

gfgit commented:
Hi, thanks for testing. I've worked on a second version of the patch plus some code style fixes detected by Clazy. I will force push soon.
All plugin are working except sysstat because library is not ported yet.

gfgit commented:
XCB not found it's because CMake components are uppercase and without `xcb-` prefix. I wonder why it worked until now...

Also there are some issues:
Ldd does not find liblxqt.so.2 so it refuses to launch panel. I had to `LD_PRELOAD` it.
I don't yet understand how this works, library is in correct path just like Qt6Xdg.so and also listed in `ldconfig -p`.
Running `sudo ldconfig` seems to rebuild the cache and fix the issue.

Another issue is panel configuration not getting installed, resulting in an empty transparent panel.
Need to investigate. For now I've manually copied it from repository to home folder

stefonarch commented:
> All plugin are working except sysstat because library is not ported yet.

That's nice to hear :)

gfgit commented:
I didn't need QTextCodec from Qt5Compat module. When I'll push, plaese check again if it's needed

tsujan commented:
> I didn't need QTextCodec from Qt5Compat module.

Just as a general note, and not a review:

Forget about Qt5Compat everywhere. Legacy encodings aren't supported by Qt6, and so, we don't have a reason to support them by force.

stefonarch commented:
I needed first https://github.com/lxqt/libsysstat/pull/47 and change  dbusmenu-qt6 → dbusmenu-lxqt to start compilation, but I'm stuck atm at `fatal error: lxqt-globalkeys.h: No such file or directory` for several plugins. 
When I change it to 
`#include <lxqt-globalkeys/lxqt-globalkeys.h>` it goes to `LXQtGlobalKeys: no such file or...` and when I include that again not found, but it's all present in `/usr/local/include`.
`

gfgit commented:
Ah yes, this is because QDebug operators are not exported. I've fixed it but forgot to push

gfgit commented:
@stefonarch I've pushed fix in [liblxqt#337](https://github.com/lxqt/liblxqt/pull/337)

gfgit commented:
Ah yes, this is because QDebug operators are not exported. I've fixed it but forgot to push

> Another issue is panel configuration not getting installed, resulting in an empty transparent panel.
> Need to investigate. For now I've manually copied it from repository to home folder

I think this is because I'm installing to `/usr/local`
The file is there but panel does not consider this path

stefonarch commented:
Compiles perfectly fine now without any warning, except that I had to disable statusnotifier plugin as I didn't manage to include `dbusmenu-lxqt` anymore, not sure if I changed something as I hit a  `git stash` by mistake. 

Panel conf is installed but not loaded at first run, don't remember how this is supposed to wirk

```
-- Installing: /usr/local/bin/lxqt-panel
-- Set non-toolchain portion of runtime path of "/usr/local/bin/lxqt-panel" to ""
-- Installing: /usr/local/share/lxqt/panel.conf
-- Installing: /usr/local/include/lxqt/lxqtpanelglobals.h
```

gfgit commented:
What's the difference between `dbusmenu-qt` and `dbusmenu-lxqt`?

I can build it with neon package for dbusmenu qt6 dev

tsujan commented:
First, `libdbusmenu-qt6` couldn't be compiled anymore. Second, it didn't exist in all distros (Arch used a self-made patch). So, we got rid of all problems by forking it (as KDE did years ago). In this way, we could fix its probable bugs too.

The Qt6 brannch of `lxqt-qtplugin` already depends on `libdbusmenu-lxqt`.

stefonarch commented:
I'll retry to include it, but really no clue atm why it didn't work anymore.

stefonarch commented:
Found my error, forgot the second part here `#include <dbusmenu-lxqt/dbusmenuimporter.h>`.

stefonarch commented:
Just found out in a nested labwc window that the qt6-panel won't work  on wayland:
![screen_area_gio_13:53:45_](https://github.com/lxqt/lxqt-panel/assets/10681413/7b2d6796-d133-4ee3-994a-5a4985d4bf14)

gfgit commented:
> Just found out in a nested labwc window that the qt6-panel won't work on wayland: ![screen_area_gio_13:53:45_](https://private-user-images.githubusercontent.com/10681413/305079454-7b2d6796-d133-4ee3-994a-5a4985d4bf14.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwMDI1MDUsIm5iZiI6MTcwODAwMjIwNSwicGF0aCI6Ii8xMDY4MTQxMy8zMDUwNzk0NTQtN2IyZDY3OTYtZDEzMy00ZWUzLTk5NGEtNWE0OTg1ZDRiZjE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE1VDEzMDMyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3ZWJmZmRlMWY3ZmUwY2MxMjRjZGIzNTk5MWZiNzczODNiYjk0OWE5NDhkNzAyNDY2ZGZkOTgyNDEzMDM3ZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2Z--FFetSkXEad8liRpzJNrLtrtcPVRkhwROqN0jhUM)

Well of course KX11Extras does not work on wayland and we use this library. As a workaround force it to use xwayland. Taskbar will be empty

tsujan commented:
Haven't we already disabled X11-dependent plugins under Wayland? If not, we should do it.

stefonarch commented:
> Well of course KX11Extras does not work on wayland and we use this library. As a workaround force it to use xwayland. Taskbar will be empty

I use without issues the qt5 version natively, so something has changed there in KF6 now.

stefonarch commented:
![screen_area_gio_14:23:53_](https://github.com/lxqt/lxqt-panel/assets/10681413/4938f4be-0e05-4873-bd42-a9638061294e)

all files which use it.

tsujan commented:
If KF6 doesn't exclude them, we could do it. Not a theoretical hassle; just a boring job.

gfgit commented:
Well I have added some assert on X11 connection somwhere. It was just for testing but it's also correct because we should exclude X11 specific parts on Wayland

tsujan commented:
I may be wrong (being busy with other codes), but I think KF5 did it for its methods.

stefonarch commented:
Looking at KX11Extras it looks (for the panel)  some functions on wayland  should use foreing-toplevel-protocol, like `static QList<WId> windows();` and others, if I got it right. Other functions (number of desktops and related) are just not available as protocol yet.

Running the panel in xwayland mode is nice to see and working fine now on labwc, but all apps launched by the menu inherit xwayland too.

stefonarch commented:
Rechecked for good and saw the  panel opens, so those are just warnings not blockers.

stefonarch commented:
After recent updates of KF6 I had to recompile it. I see (also before) this:

```
$ lxqt-panel
starting
WinIdChange 1600009 handle QWidgetWindow(0x56007fda3e90, name="LXQtPanel panel1Window") QScreen(0x56007fa91b30, name="Virtual1")
StatusNotifierProxy, services: QList()
Manager selection claimed
trying to dock window  18874385
adding damage watch for  18874385
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.Menu was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.IconThemePath was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.ToolTip was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.OverlayIconName was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.IconName was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.AttentionIconName was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.IconThemePath was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.ToolTip was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.OverlayIconPixmap was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.AttentionIconPixmap was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.IconName was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.IconThemePath was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.ToolTip was not found in object /StatusNotifierItem)
Error on DBus request(:1.71,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.UnknownProperty, Property org.kde.StatusNotifierItem.IconName was not found in object /StatusNotifierItem)


```

gfgit commented:
Ops, sorry I edited your post instead of replying.
I had those same DBus errors in Qt5 branch

stefonarch commented:
> I had those same DBus errors in Qt5 branch

Strange, here I've only some things about icons:

```
$ lxqt-panel
WinIdChange 1 handle QWidgetWindow(0x64309003bd20, name="LXQtPanel panel1Window") QScreen(0x64308fe9fd60, name="eDP-1")
WinIdChange 2 handle QWidgetWindow(0x6430904091b0, name="LXQtPanel panel_{2aaae4ef-94e6-450c-a8d6-7f862d156da2}Window") QScreen(0x64308fe9fd60, name="eDP-1")
StatusNotifierProxy, services: ()
Error on DBus request(:1.132,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.Failed, error occurred in Get)
Error on DBus request(:1.132,/StatusNotifierItem): QDBusError(org.freedesktop.DBus.Error.Failed, error occurred in Get)
Error on DBus request(:1.94,/org/ayatana/NotificationItem/gammastep): QDBusError(org.freedesktop.DBus.Error.InvalidArgs, Proprietà «ToolTip» inesistente)
Error on DBus request(:1.94,/org/ayatana/NotificationItem/gammastep): QDBusError(org.freedesktop.DBus.Error.InvalidArgs, Proprietà «OverlayIconName» inesistente)
Error on DBus request(:1.94,/org/ayatana/NotificationItem/gammastep): QDBusError(org.freedesktop.DBus.Error.InvalidArgs, Proprietà «OverlayIconPixmap» inesistente)
Error on DBus request(:1.94,/org/ayatana/NotificationItem/gammastep): QDBusError(org.freedesktop.DBus.Error.InvalidArgs, Proprietà «AttentionIconPixmap» inesistente)
```

gfgit commented:
In the meantime I went ahead a bit :)

Rullo di tamburi...

![lxqt-panel-kde-plasma-wayland](https://github.com/lxqt/lxqt-panel/assets/4284724/bc502e10-0d7d-44b9-9940-7a881317eb23)

So this is LXQt Panel running natively on KDE Plasma Wayland session!

How to achieve this:
- Qt6 branch
- #2007 rebased on top of this
- `LayerShellQt` integration for placing panel on bottom part of screen
- `org_kde_plasma_window_management` protocol to manage windows on KWin
- Some black magic of course

tsujan commented:
> Some black magic of course

I like it so much when you do your black magic. Well done!

stefonarch commented:
Wow, senza parole :)

>LayerShellQt integration for placing panel on bottom part of screen

Which version?  No fullscreen popups ?

gfgit commented:
My long term idea is:
- Remove all Qt5 related hacks and workaround (I've seen many of them)
- Move all window-system related code to a single class (used by auto-hide logic, show desktop, taskbar, desktop switch, etc)
- Split X11/Wayland at compile time (just to enforce clean code, this way it's easy to see misuse of X11 API on Wayland)
- Choose appropriate backend at runtime: xcb or Wayland

On Wayland there isn't yet a standard, so the backend would have sub-backends which are using some specific protocol or even using DBus (example for Sway).
Waybar is a good example: it supports various compositors and chooses the correct communication protocol by probing at start.

For now I've added support for KWin using KDE's in-house protocol. I think this is the first non-KDE software using their window managment protocol.
I've copied and slightly adapted some code from [libtaskmanager/waylandtasksmodel.cpp](https://invent.kde.org/plasma/plasma-workspace/-/blob/master/libtaskmanager/waylandtasksmodel.cpp) from Plasma Workspace repository

tsujan commented:
At some point, you'll setup a LabWC environment and see how comfortably you could work there.

Anyway, with your works, I'm not worried about the future of LXQt+Wayland anymore :)

gfgit commented:
Quick update:
Now reserving screen area works and also placing panel on top/bottom.
Right/Left is a bit awkward because it does not count Plasma panel occupied area so the panel is too tall and gets cutted.
But this would not happen in a LXQt session without other programs claiming screen edges with layer shell.
Still it could be improved if possible.

LayerShellQt is at version 6.0.80 which allows popups to be normal XDG Popup windows, if that was your question.

stefonarch commented:
> LayerShellQt is at version 6.0.80 which allows popups to be normal XDG Popup windows, if that was your question.

Thanks, I don't have it yet in Neon.

gfgit commented:
Now also `plugin-sysstat` is ported to Qt6!

gfgit commented:
```
lxqt-panel: error while loading shared libraries: libdbusmenu-lxqt.so.0: cannot open shared object file: No such file or directory
```

Similar LDD issue fixed by running `sudo ldconfig`

tsujan commented:
Not a review yet (no stable KF6 here), but, in general:

When porting components to Qt6, it's preferable to change `QVector` to `QList` everywhere because the former is just an alias for the latter.

I think I've done so in my Qt6 branches; if not, please tell me.

tsujan commented on panel/config/addplugindialog.cpp:
```diff
@@ -34,10 +34,13 @@
+#include <QString>
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Any reason for using `chrono`? `QTimer::setInterval(int msec)` exists and `QTimer::setInterval(std::chrono::milliseconds value)` is an overloaded function.

Apart from thai, IMHO, it's highly preferable not to change a code while porting, unless really necessary.

gfgit replied:
Hi, you don't need to search all references. It's all in commit b5474baff73f6d88bb63b4d45f6975efcbb5731a

It was automatically done by Clazy and I think this way it's far more readable, but I can move the commit is a separate PR after Qt6 port.

tsujan replied:
> but I can move the commit in a separate PR

Let me explain it pragmatically, from a reviewer's perspective; it might convince you if you put yourself in my shoes.

Porting to a new version of Qt may seem straightforward in most places, but (1) the amount of changes is heavy, and (2) there's always the danger of introducing regressions, especially where the new version of Qt shows an undocumented different behavior (which has happened  in my port of libfm-qt). To avoid such regressions as much as possible, we need to stick with the current code during porting and change them only if *really* needed, such that **the changed codes reflect what is really needed to be done** (there are several examples of that in https://github.com/lxqt/libfm-qt/pull/778/files).

As for a component like libfm-qt or pcmanfm-qt, I've used, tested and fixed its Qt6 port continuously for more than a year. Sadly, we won't have that time for the remaining components. Hence, the less unnecessary changes, the better.

tsujan commented on panel/lxqtpanelapplication.cpp:
```diff
@@ -25,17 +25,19 @@
+#include <LXQt/Settings>
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Ditto.

tsujan commented on plugin-desktopswitch/desktopswitchconfiguration.cpp:
```diff
@@ -25,9 +25,12 @@
+#include <KX11Extras>
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Ditto.

tsujan commented on plugin-fancymenu/lxqtfancymenu.cpp:
```diff
@@ -49,6 +49,9 @@
 #include <QDir>
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Ditto.

tsujan commented on plugin-fancymenu/lxqtfancymenuwindow.cpp:
```diff
@@ -60,6 +60,9 @@
 #include <QStyledItemDelegate>
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Again.

tsujan commented on plugin-mainmenu/lxqtmainmenu.cpp:
```diff
@@ -50,10 +50,14 @@
+#include <XdgAction>
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Again.

tsujan commented on plugin-mount/actions/deviceaction_menu.cpp:
```diff
@@ -28,14 +28,17 @@
 #include "../popup.h"
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Ditto.

tsujan commented on plugin-sensors/lxqtsensors.cpp:
```diff
@@ -26,13 +26,16 @@
+#include "lxqtsensorsconfiguration.h"
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Ditto.

tsujan commented on plugin-statusnotifier/statusnotifierbutton.cpp:
```diff
@@ -30,10 +30,13 @@
+#include <chrono>
+#include <dbusmenu-qt6/dbusmenuimporter.h>
+
+using namespace std::chrono_literals;
```
Ditto.

tsujan commented on plugin-statusnotifier/statusnotifierwidget.cpp:
```diff
@@ -27,9 +27,12 @@
+#include "statusnotifierproxy.h"
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Ditto.

tsujan commented on plugin-taskbar/lxqtgrouppopup.cpp:
```diff
@@ -29,13 +29,18 @@
-#include <QDebug>
+#include <chrono>
+
+using namespace std::chrono_literals;
```
Again

tsujan commented on plugin-taskbar/lxqttaskbutton.cpp:
```diff
@@ -47,17 +47,22 @@
+#include <X11/Xlib.h>
+#include <chrono>
+
+using namespace std::chrono_literals;
```
I guess there remain lots of other instances too....

gfgit commented:
@tsujan I've ported away from `QVector` and `QScopedPointer`, commit about `chrono` is removed, will put it in a future PR

stefonarch commented:
Basically here can be seen 2 issues:

* statusnotifier items menus not working
* localization missing of month/day text in wordclock, I noticed it was already present in Neon.

romangg commented on plugin-qeyes/qeyesimagewidget.cpp:
```diff
@@ -32,16 +32,21 @@
-    } else {
+    }
+    else
+    {
```
Are these style changes intentional?

gfgit replied:
Personally I think putting braces on newline makes it more readable.
We don't have clang format so style is not consistent.

romangg commented on plugin-tray/fdoselectionmanager.cpp:
```diff
@@ -127,7 +138,7 @@ bool FdoSelectionManager::addDamageWatch(xcb_window_t client)
 }
 
-bool FdoSelectionManager::nativeEventFilter(const QByteArray &eventType, void *message, long int *result)
+bool FdoSelectionManager::nativeEventFilter(const QByteArray &eventType, void *message, qintptr *result)
```
Simply int works too?

gfgit replied:
Well I've copied Qt signature inside `QAbstractNativeEventFilter`.
I don't have a strong opinion about using Qt typedefs

romangg commented on plugin-tray/sniproxy.cpp:
```diff
@@ -580,20 +582,23 @@ void SNIProxy::sendClick(uint8_t mouseButton, int x, int y)
         event->response_type = XCB_BUTTON_PRESS;
         event->event = m_windowId;
-        event->time = QX11Info::getTimestamp();
+        event->time = XCB_CURRENT_TIME; //FIXME: how to get proper timestamp?
```
No need for comment. Setting XCB_CURRENT_TIME should be fine.

gfgit replied:
In following PRs  I've linked to Qt private API which was needed for KDE Window managment protocol requiring wl_surface.
So I've also changed this to use private platform integration resources

romangg reviewed (commented):
Looks good overall to me. I would try to replace all the new Xlib includes instead with xcb includes.

gfgit commented on panel/panelpluginsmodel.h:
```diff
@@ -336,6 +336,6 @@ public slots:
 };
 
-Q_DECLARE_METATYPE(Plugin const *)
+Q_DECLARE_OPAQUE_POINTER(Plugin const *)
```
It gave me compile error because Plugin is forward declared. I should check this again

gfgit replied:
Better yet, I've moved the macro to `plugin.h` where class is actually defined.
Could this have side effects?

romangg replied:
Don't think so.

gfgit commented on plugin-colorpicker/colorpicker.cpp:
```diff
@@ -143,12 +142,22 @@ void ColorPickerWidget::mouseReleaseEvent(QMouseEvent *event)
+    }
+    else
+    {
+        qWarning() << "WAYLAND does not support grabbing windows";
```
I've already implemented it in #2031 ;-)

gfgit commented on plugin-colorpicker/colorpicker.cpp:
```diff
@@ -36,6 +36,10 @@
 #include <QSvgRenderer>
 
+//NOTE: Xlib.h defines Bool which conflicts with QJsonValue::Type enum
+#include <X11/Xlib.h>
```
If I understand your code it'always first/last screen depending on traverse order?

romangg replied:
Not necessarily, it's the iterator where `screen == 0`.  While `screen` is initialized with some value `int screen = data.screen_number;` and decreases on every iterator step.

gfgit replied:
Ok, then where is `data.screen_number` coming from?

romangg replied:
Depends on the platform:
* [X11](https://github.com/winft/como/blob/b7a040a39d4785c25810ae0d1f90703078d787b7/como/base/x11/platform.h#L65)
* [Xwayland](https://github.com/winft/como/blob/b7a040a39d4785c25810ae0d1f90703078d787b7/como/xwl/xwayland.h#L240)

So rather simple. :smile:

gfgit replied:
Maybe it's `x11_data.screen_number = QX11Info::appScreen();` ?
But this also needs private Qt API

stefonarch commented:
Worldclock issue: beside the non localized month/day names in the widget I noticed also that timezones can't be set anymore.
![screen_area_ven_17:05:30_](https://github.com/lxqt/lxqt-panel/assets/10681413/4d8ac618-7aaf-41a4-8c03-d7bdc9db547c)

stefonarch commented on plugin-worldclock/lxqtworldclockconfigurationtimezones.cpp:
```diff
@@ -119,7 +119,7 @@ int LXQtWorldClockConfigurationTimeZones::updateAndExec()
             qStrings.prepend(tr("Other"));
 
-        QTreeWidgetItem *tzItem = new QTreeWidgetItem(QStringList() << qStrings[qStrings.length() - 1] << timeZone.displayName(now) << timeZone.comment() << QLocale::countryToString(timeZone.country()));
+        QTreeWidgetItem *tzItem = new QTreeWidgetItem(QStringList() << qStrings[qStrings.length() - 1] << timeZone.displayName(now) << timeZone.comment() << QLocale::territoryToString(timeZone.territory()));
```
My understanding is that failing localization and timezone settings are caused by this change to `territory`.
Deprecated:

>Returns a [QString](https://doc.qt.io/qt-6/qstring.html) containing the name of country.

> Returns the two-letter territory code for territory, as defined in the ISO 3166 standard.

tsujan commented on plugin-sysstat/lxqtsysstatutils.cpp:
```diff
@@ -44,13 +44,14 @@ QString netSpeedToString(int value)
 }
 
-int netSpeedFromString(QString value)
+int netSpeedFromString(const QStringView& value)
```
[Qt doc](https://doc.qt.io/qt-6/qstringview.html#details) says, "QStringViews should be passed by value, not by reference-to-const:". Would you please change this?

gfgit replied:
Done

tsujan replied:
Very good.

tsujan commented on plugin-worldclock/lxqtworldclock.cpp:
```diff
@@ -165,6 +164,8 @@ void LXQtWorldClock::restartTimer()
 
 void LXQtWorldClock::settingsChanged()
 {
+    static QRegularExpression regexp(QLatin1String("'[^']*'"));
```
`static const` would be better.

tsujan commented on plugin-worldclock/lxqtworldclock.cpp:
```diff
@@ -476,7 +478,8 @@ void LXQtWorldClock::updatePopupContent()
 bool LXQtWorldClock::formatHasTimeZone(QString format)
 {
-    format.replace(QRegExp(QLatin1String("'[^']*'")), QString());
+    static QRegularExpression regexp(QLatin1String("'[^']*'"));
```
I agree; like the previous instance.

gfgit replied:
Done. Could you check new syntax is correct? Also in other places...

tsujan replied:
It's OK.

tsujan commented:
> Worldclock issue: beside the non localized month/day names in the widget I noticed also that timezones can't be set anymore.

Found the cause of the first, which is most probably the cause of the second. It isn't about a bug but a change of behavior in Qt6. Will make a PR after merging Qt6 PRs because it deserves a separate PR.

@gfgit, if you make the requested changes, I could do the merging today.

tsujan commented on plugin-taskbar/lxqttaskgroup.cpp:
```diff
@@ -657,18 +660,29 @@ bool LXQtTaskGroup::onWindowChanged(WId window, NET::Properties prop, NET::Prope
+                std::for_each(buttons.begin(), buttons.end(), std::bind(&LXQtTaskButton::setUrgencyHint, std::placeholders::_1, urgency || info.hasState(NET::DemandsAttention)));
+                set_urgency = false;
+            }
+
```
`set_urgency = false;` inside `if (!set_urgency)`?

Making such changes in a PR that isn't about them will cause complications.

@palinek 
Would you please look into this? The code is yours, I think.

tsujan replied:
@gfgit 
`set_urgency = false` had a correct place before. You also changed the place where `urgency` was assigned a few lines above. Would you make the code as it was before (of course, except for using `nativeInterface`)?

gfgit replied:
Original logic was badly written. Anyway this code will go away when moving to WM Abstraction, so we can keep it for now.

tsujan replied:
I'm afraid, this isn't still the original code

> Original logic was badly written

Even if so, we should not make such changes in this PR; otherwise, it'll be very difficult to find causes of problems. For having a clean and searchable history, each job must be done in a separate PR.

Please reverse all changes here, except for using `nativeInterface`!

gfgit replied:
You realize this is quite silly? Moving 2 lines in a common place is far more readable than squashing 10 unrelated commits together...

tsujan replied:
> You realize this is quite silly?

On the contrary, now we know that this PR won't change anything in this regard after being merged. So, if some problem exists, we won't waste our time. That's how I found the cause of localization problem in Worldclock (https://github.com/lxqt/lxqt-panel/pull/2024#issuecomment-2009619729).

tsujan commented on plugin-tray/xcbutils.h:
```diff
@@ -53,13 +52,19 @@ namespace Xcb
+    static inline void cleanup(void *pointer) noexcept { free(pointer); }
+    void operator()(void *pointer) const noexcept { cleanup(pointer); }
+};
+
```
@palinek, please also take a look at this (your code again).

gfgit replied:
What about this?

gfgit replied:
I took it from Qt by the way: [qscopedpointer.h#L51](https://github.com/qt/qtbase/blob/ceeaf1b65787d3ef46dd0bdbd0e6810137163fea/src/corelib/tools/qscopedpointer.h#L51)

gfgit replied:
Actually `cleanup()` seems to be called only by `QScoperPointer` and since we ported code to `std::unique_ptr` we can get rid of it and free directly.

tsujan replied:
I had no problem with this one; just asked @palinek to take a look at it, for the sake of certainty.

tsujan commented on plugin-taskbar/lxqttaskbutton.cpp:
```diff
@@ -395,7 +421,8 @@ void LXQtTaskButton::mouseMoveEvent(QMouseEvent* event)
     // release mouse appropriately, by positioning the event outside
     // the button rectangle (otherwise, the button will be toggled)
-    QMouseEvent releasingEvent(QEvent::MouseButtonRelease, QPoint(-1,-1), Qt::LeftButton, Qt::NoButton, Qt::NoModifier);
+    //FIXME: constructor is deprecated, also seems a bit hacky code
```
This comment isn't needed; your code is correct. Please remove the comment.

gfgit replied:
Still the code is a bit hacky and could be reworked. Don't know yet how.

tsujan replied:
This of a code reader who might encounter "//FIXME: constructor is deprecated"  and ask, "which deprecated c-tor? This isn't deprecated."  I mean it'll be confusing later.

gfgit commented:
I suggest this time to NOT squash the commit before merging.
It would result in an commit too big too be read.
I've carefully divided commits to make changes easily understandable and this could be useful as a reference
for other people wanting to port their software to Qt 6.
Some commits name have "TODO" in their messages, this of course will be removed if changes.
Too reduce the number of commits some of them can be grouped based on topic.

gfgit commented:
CI doesn't find Qt6 dependencies

tsujan commented:
> I suggest this time to NOT squash the commit before merging.

OK.

> CI doesn't find Qt6 dependencies

No worries; we've got used to CI's temporary issues and tolerate them.

Will merge all Qt6-related PRs in the proper order tomorrow.

tsujan commented:
PRs should be squashed before being merged. A clean history is a must. I just agreed that you don't do it yourself.

As for code readers, it's impossible to make everything straightforward for them; they need to delve into codes.

gfgit commented:
> PRs should be squashed before being merged. A clean history is a must. I just agreed that you don't do it yourself.
> 
> As for code readers, it's impossible to make everything straightforward for them; they need to delve into codes.

Ok for clean history but at least squashing in groups leading to say 3 final commits instead to a single one coild have been a good trade off

tsujan commented:
@gfgit 
It wasn't my personal decision. I did it according to a sort of tradition that has existed in LXQt for a long time (I got familiar to it after joining LXQt years ago). Some devs even believe that the PR maker himself should squash his commits, because a PR is about a single change.
